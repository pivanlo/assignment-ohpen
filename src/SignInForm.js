import React from 'react';
import ReactDOM from 'react-dom';
import { Link, Redirect } from "react-router-dom";
import Button from 'react-bootstrap/Button';
import Form from 'react-bootstrap/Form';
import { Auth } from "aws-amplify";

class SignInForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      username: "",
      password: "",
      formError: ""
    };

    this.handleUsernameChange = this.handleUsernameChange.bind(this);
    this.handlePasswordChange = this.handlePasswordChange.bind(this);
    this.clearFormError = this.clearFormError.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleUsernameChange(event) {
    this.setState({ username: event.target.value });
  }

  handlePasswordChange(event) {
    this.setState({ password: event.target.value });
  }

  clearFormError(event) {
    this.setState({ formError: null });
  }

  async handleSubmit(event) {
    event.preventDefault();

    const username = this.state.username;
    const password = this.state.password;

    const formError = this.validateForm(username, password);

    if (formError) {
      return;
    }

    // We will need the username later if the user still
    // needs to be verified.
    this.props.auth.setUsername(username);

    try {
      const user = await Auth.signIn(
        this.state.username,
        this.state.password
      );

      this.props.auth.setAuthStatus(true);
      this.props.auth.setUser(user);

    } catch (error) {
      if (error.code === "UserNotConfirmedException") {
        this.props.history.push("/verify");
      } else {
        this.setState({ formError: 'Incorrect username or password.' });
      }
    }
  }

  validateForm(username, password) {
    let message;

    if (!username) {
      message = 'The username cannot be empty.';
    } else if (!password) {
      message = 'The password cannot be empty.';
    }

    this.setState({ formError: message });
    return message;
  }

  render() {
    if (this.props.auth.isAuthenticated && this.props.auth.user) {
      return <Redirect to='/profile' />;
    }

    return (
      <div className="formContainer">
        <Form className="form-signin" onSubmit={this.handleSubmit}>
          <h1>Sign in</h1>

          <Form.Group controlId="formGroupUsername">
            <Form.Label>Username</Form.Label>
            <Form.Control type="text" value={this.state.username}
              onChange={this.handleUsernameChange}
              onFocus={this.clearFormError}
            />
          </Form.Group>

          <Form.Group controlId="formGroupPassword">
            <Form.Label>Password</Form.Label>
            <Form.Control type="password" value={this.state.password}
              onChange={this.handlePasswordChange}
              onFocus={this.clearFormError}
            />
          </Form.Group>

          <Button type="submit" variant="dark" block>Sign in</Button>

          {!this.state.formError &&
            <div className="mt-3">
              <span>I'm a new user, </span>
              <Link to="/sign-up">Sign up</Link>
            </div>
          }

          {this.state.formError &&
            <div className="mt-3">
              <b>{this.state.formError}</b>
            </div>
          }
        </Form>
      </div>
    );
  }
}

export default SignInForm;
