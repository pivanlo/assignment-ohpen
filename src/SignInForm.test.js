import React from 'react';
import SignInForm from "./SignInForm.js";
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';

it('renders correctly for unauthenticated users', () => {
  const auth = {
    user: {
      attributes: {
        email: 'john.doe@gmail.com'
      }
    },
    isAuthenticated: false,
    isManager: () => false
  };

  const tree = renderer
    .create(
      <BrowserRouter>
        <SignInForm auth={auth} />
      </BrowserRouter>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders correctly for authenticated users', () => {
  const auth = {
    user: {
      attributes: {
        email: 'john.doe@gmail.com'
      }
    },
    isAuthenticated: true,
    isManager: () => false
  };

  const tree = renderer
    .create(
      <BrowserRouter>
        <SignInForm auth={auth} />
      </BrowserRouter>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
