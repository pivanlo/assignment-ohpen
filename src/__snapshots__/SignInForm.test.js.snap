// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly for authenticated users 1`] = `null`;

exports[`renders correctly for unauthenticated users 1`] = `
<div
  className="formContainer"
>
  <form
    className="form-signin"
    onSubmit={[Function]}
  >
    <h1>
      Sign in
    </h1>
    <div
      className="form-group"
    >
      <label
        className="form-label"
        htmlFor="formGroupUsername"
      >
        Username
      </label>
      <input
        className="form-control"
        id="formGroupUsername"
        onChange={[Function]}
        onFocus={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label
        className="form-label"
        htmlFor="formGroupPassword"
      >
        Password
      </label>
      <input
        className="form-control"
        id="formGroupPassword"
        onChange={[Function]}
        onFocus={[Function]}
        type="password"
        value=""
      />
    </div>
    <button
      className="btn btn-dark btn-block"
      disabled={false}
      type="submit"
    >
      Sign in
    </button>
    <div
      className="mt-3"
    >
      <span>
        I'm a new user, 
      </span>
      <a
        href="/sign-up"
        onClick={[Function]}
      >
        Sign up
      </a>
    </div>
    
  </form>
</div>
`;
